www/app/blog/2024-11-19-distributed-ddl.mdx

```diff
@@ -0,0 +1,126 @@
+
+In order to intercept the DDL before it executes, `nile_ddl` uses the `processUtility_hook`. To understand why and how our extension uses this hook, we first need to explain what a utility command is and how PostgreSQL handles them. 
+
+In Postgres terms, a "utility" is any command except `SELECT`, `INSERT`, `UPDATE` and `DELETE`. This includes all DDL commands as well as other commands like `COMMIT`, `NOTIFY` or `DO`. When Postgres recieves a utility command, it uses `ProcessUtility(..)` method to process it. This is a simple wrapper that looks like this:
+
+```c
+	if (ProcessUtility_hook)
+		(*ProcessUtility_hook) (...);
+	else
+		standard_ProcessUtility(...);
+```
+
+This method checks if there are any extensions that want to process the utility command before Postgres runs its standard processing. The `nile_ddl` extension provides such a hook and, as a result, is called before Postgres processes any command. This is incredibly useful because `processUtility_hook` is triggered for nearly everything that isn't a query or a DML, giving us a single entry point for almost everything we need to handle. 
+
+Once we are finish processing the command, it is our responsibility to call `standard_ProcessUtility`,  so that Postgres can continue its normal flow. A bit off-topic, but in case you're curious: `standard_ProcessUtility` method is essentially a gigantic switch statement that routes every one of the 60+ utility commands to their appropriate handler.
+
+![Process utility hook](/blog/2024-11-19-distributed-ddl/process_utility_hook.png)
+
+So, our extension gets called with a utility command. What happens next? At this point, it needs to perform a few tasks:
+
+1. **Check that the command is actually a DDL.** If the command isn't `ALTER`, `CREATE` or `DROP` we can skip it.
+2. **Raise error for unsupported DDL.** Nile has specific restrictions on the type of relations we allow. For example, primary keys **must** include the `tenant_id` column, and `tablespace` commands are not supported at all (Nile handles these automatically). 
+3. **Determine which locks the DDL requires.**
+4. **Begin a distributed transaction** (only needed if this is the first DDL in a transaction)
+5. **Ask the transaction coordinator to distribute the locks**
+
+The reason we extract the locks and distribute them early in the process is to minimize the time spent holding a lock. Most DDL statements require an `ACCESS EXCLUSIVE` lock, which not only prevents any queries from accessing the table in question while the DDL is executing but also prevents any new queries from accessing the table while the DDL is waiting to acquire the lock. To minimize the risk and the time spent while holding the lock, Nile attempts to acquire the necessary locks - with a short lock timeout - on all relevant databases before starting to execute the DDLs. If the lock acquisition fails on any database, the DDL will return an error rather than continue waiting for the lock.
```

Comment:
the main reason I think is to be able to distribute the locks in a predictable order (at least when possible), and before executing the queries. the minimises the risk of deadlocks and other lock conflicts, which is bad for expensive DDL (eg create index) which could be failing later on on one of the nodes while succeeding in others.